ea-2022-04-ndvi-automation.ipynb

```diff
@@ -0,0 +1,1600 @@
+    "    for i, aband in enumerate(paths):\n",
+    "\n",
+    "        cleaned_band = rxr.open_rasterio(aband,\n",
+    "                                         masked=True).rio.clip(clip_bound,\n",
+    "                                                               from_disk=True).squeeze()\n",
+    "        # Clean the data\n",
+    "        valid_range = (0, 10000)\n",
+    "        # Only run this step if a valid range tuple is provided\n",
+    "        if valid_range:\n",
+    "            mask = ((cleaned_band < valid_range[0]) | (\n",
+    "                cleaned_band > valid_range[1]))\n",
+    "            cleaned_band = cleaned_band.where(\n",
+    "                ~xr.where(mask, True, False))\n",
+    "\n",
+    "        cleaned_band[\"band\"] = i+1\n",
+    "        all_bands.append(cleaned_band)\n",
+    "\n",
+    "    if stack:\n",
+    "        return xr.concat(all_bands, dim=\"band\")\n",
+    "    else:\n",
+    "        print(\"Returning a list of xarray objects.\")\n",
+    "        return all_bands\n",
+    "\n",
+    "\n",
+    "def remove_clouds(path_to_qa):\n",
+    "    \"\"\"Opens landsat qa file, selects cloud/shadow values, creates cloud mask.\n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    path_to_qa : string\n",
+    "        Path to pixel_qa file in landsat data that will be used to clean data\n",
+    "        for presence of clouds and shadows.\n",
+    "\n",
+    "    Returns\n",
+    "    ----------\n",
+    "    cl_mask : xarray object\n",
+    "        The cloud mask xarray object that will be used to clean landsat data.\n",
+    "        \"\"\"\n",
+    "\n",
+    "    landsat_qa = rxr.open_rasterio(path_to_qa).squeeze()\n",
+    "\n",
+    "    high_cloud_confidence = em.pixel_flags[\"pixel_qa\"][\"L8\"][\"High Cloud Confidence\"]\n",
+    "    cloud = em.pixel_flags[\"pixel_qa\"][\"L8\"][\"Cloud\"]\n",
+    "    cloud_shadow = em.pixel_flags[\"pixel_qa\"][\"L8\"][\"Cloud Shadow\"]\n",
+    "\n",
+    "    all_masked_values = cloud_shadow + cloud + high_cloud_confidence\n",
+    "    cl_mask = landsat_qa.isin(all_masked_values)\n",
+    "\n",
+    "    return cl_mask\n",
+    "\n",
+    "\n",
+    "def combine_scenes(path_to_all_scenes, crop_path):\n",
+    "    \"\"\"Open, cleans, crops list of raster files, cleans data with cloud mask, \n",
+    "    captures average ndvi, date, and site name and saves to pandas df. \n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    path_to_all_scenes: string\n",
+    "        Path to directory with all landsat scenes/date folders to be combined.\n",
+    "    crop_path: string\n",
+    "        A string path to the geodataframe containing the crop geometry that \n",
+    "        you wish to crop your data to.\n",
+    "\n",
+    "    Returns\n",
+    "    ----------\n",
+    "    df : pandas dataframe\n",
+    "        Dataframe containing average NDVI values, date, and site name. \n",
+    "        \"\"\"\n",
```

Comment:
Great pseudocode for all of your functions. Easy to read and understand what each function does!

Reply:
Thank you!